source/event.d

```diff
@@ -1,51 +1,360 @@
+ *     foo.click(4,5); // callback!
+ * }
+ * --------------------
+ */
+mixin template genCallback(T, E : Event, args ...)
+{
+	static assert(is(T == class));
+
+	/**
+	 * if parameters as passed then use them as the callback delegate parameters,
+	 * otherwise just pass the event itself
+	 */
+	static if (args.length)
+	{
+		import std.meta : AliasSeq, Filter, templateNot;
+		import std.traits : isType;
+
+		private alias _as = AliasSeq!args;
+		private alias _types = Filter!(isType, _as);
+		private enum _names = toEventFormat!(Filter!(templateNot!isType, _as));
+
+		static assert(Filter!(templateNot!isType, _as).length == _types.length);
+	}
+	else
+	{
+		private alias _types = const E;
+		private enum _names = ",event";
+	}
+
+public:
+	pure
+	void connect(void delegate (T, _types) dg)
+		in (dg.ptr !is null)
+	{
+		callback = dg;
+	}
+
+protected:
+	void dispach(E event)
+	{
+		import std.string : format;
+		if (callback.ptr !is null)
+			mixin("callback(this%s);".format(_names));
+	}
+
+	void delegate(T, _types) callback;
+}
+
+
+/** Generate a string with the format `,event.<args[0],event.args[1],...`
+ *
+ * This template is used **internaly** only!
+ *
+ * Params:
+ *     args = string values to be formated
+ *
+ * Examples:
+ * --------------------
+ * enum str = toEventFormat!("a", "b", "c");
+ * --------------------
+ * --------------------
+ * static assert(toEventFormat!("a", "b"), ",event.a,event.b");
+ * --------------------
+ *
+ * Returns:
+ *     string `enum`
+ */
+private template toEventFormat(args ...)
+{
+	@safe pure
+	auto toEventFormat()
+	{
+		string ret;
+		foreach (var; args)
+		{
+			import std.string : format;
+			import std.traits : isType;
+			static assert(!isType!var);
+			static assert(is(typeof(var) == string));
+			ret ~= ",event." ~ var;
+		}
+		return ret;
+	}
+}
+
+@("Event: toEventFormat") @safe
+unittest
+{
+	assertEquals(toEventFormat!("a", "b"), (",event.a,event.b"));
+}
+
+
+version(unittest) import aurorafw.unit;
+version(unittest)
+{
+	@safe pure
+	void onFooEvent(in Foo sender, in int a, in int b, in int c)
+	{
+		assertEquals([a,b,c], [2,4,5]);
+	}
+
+	@safe pure
+	void onBarEvent(in Foo sender, in BarEvent event)
+	{
+		assertEquals(event.toString, "BarEvent");
+	}
+
+
+	@EventType("FooEvent")
+	@safe class FooEvent : Event
+	{
+		mixin basicEventType!FooEvent;
+
+	public:
+		this(in int a, in int b, in int c)
+		{
+			this.a = a;
+			this.b = b;
+			this.c = c;
+		}
+
+		const int a, b, c;
+	}
 
-	pragma(inline, true)
-	override @property string name() const
+	@EventType("BarEvent")
+	@safe class BarEvent : Event
 	{
-		import std.conv : to;
-		return eventType.to!string;
+		mixin basicEventType!BarEvent;
+
+	public:
+		this(in size_t x, in size_t y)
+		{
+			this.x = x;
+			this.y = y;
+		}
+
+		const size_t x, y;
 	}
+
+	import std.typecons : scoped;
+	class Foo
+	{
+		mixin genCallback!(const Foo, FooEvent, const int, "a", const int, "b", const int, "c") onFoo;
```

Comment:
I guess you have this changed on other PR. If you do so, ignore this comment, but:

You can generate `onFoo` symbol inside the `mixin` by parsing the Event type name.

Reply:
No, `onFoo` is refering to to the mixin's scope. The mixin doesn't generate unique function names, every callback will have a `connect`, `dispach` and a `callback`delegate, because of this, you need to diferentiate between callbackA and callbackB since the functions have identical names and you can do that by naming the mixin's scope.

Reply:
Sure, you're right. Maybe it's more readable and flexible too.